File: sol/2516/2516.cpp
struct Solution {
    int takeCharacters(string s, int k) {
        int n = s.size(), totals[3] = {0};
        for (char c : s){
            ++totals[c - 'a'];
        }
        if (totals[0] < k || totals[1] < k || totals[2] < k){
            return -1;
        }
        int res = -1, l = 0, counts[3] = {0}, idx;
        for (int r = 0; r < n; ++r){
            idx = s[r] - 'a';
            ++counts[idx];
            while (l <= r && counts[idx] > totals[idx] - k){
                --counts[s[l++] - 'a'];
            }
            res = max(res, r - l + 1);
        }
        return n - res;
    }
};

File: sol/2516/2516.py
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        n = len(s)
        totals = [0] * 3
        for c in s:
            totals[ord(c) - ord('a')] += 1
        if any(x < k for x in totals):
            return -1
        res = -1
        l = 0
        counts = [0] * 3
        for r in range(n):
            idx = ord(s[r]) - ord('a')
            counts[idx] += 1
            while l <= r and counts[idx] > totals[idx] - k:
                counts[ord(s[l]) - ord('a')] -= 1
                l += 1
            res = max(res, r - l + 1)
        return n - res

